engine/consensus/matching/engine.go

```diff
@@ -149,1085 +105,209 @@ func New(
+	}
 
 	return e, nil
 }
 
-// Ready returns a ready channel that is closed once the engine has fully
-// started. For the propagation engine, we consider the engine up and running
-// upon initialization.
-func (e *Engine) Ready() <-chan struct{} {
-	return e.unit.Ready()
-}
-
-// Done returns a done channel that is closed once the engine has fully stopped.
-// For the propagation engine, it closes the channel when all submit goroutines
-// have ended.
-func (e *Engine) Done() <-chan struct{} {
-	return e.unit.Done()
-}
-
-// SubmitLocal submits an event originating on the local node.
-func (e *Engine) SubmitLocal(event interface{}) {
-	e.Submit(e.me.NodeID(), event)
-}
-
-// Submit submits the given event from the node with the given origin ID
-// for processing in a non-blocking manner. It returns instantly and logs
-// a potential processing error internally when done.
-func (e *Engine) Submit(originID flow.Identifier, event interface{}) {
-	e.unit.Launch(func() {
-		err := e.Process(originID, event)
-		if err != nil {
-			engine.LogError(e.log, err)
-		}
-	})
-}
-
-// ProcessLocal processes an event originating on the local node.
-func (e *Engine) ProcessLocal(event interface{}) error {
-	return e.Process(e.me.NodeID(), event)
-}
-
-// Process processes the given event from the node with the given origin ID in
-// a blocking manner. It returns the potential processing error when done.
+// Process sends event into channel with pending events. Generally speaking shouldn't lock for too long.
 func (e *Engine) Process(originID flow.Identifier, event interface{}) error {
-	return e.unit.Do(func() error {
-		return e.process(originID, event)
-	})
+	e.pendingEventSink <- &Event{
+		OriginID: originID,
+		Msg:      event,
+	}
+	return nil
 }
 
-// HandleReceipt pipes explicitly requested receipts to the process function.
-// Receipts can come from this function or the receipt provider setup in the
-// engine constructor.
-func (e *Engine) HandleReceipt(originID flow.Identifier, receipt flow.Entity) {
-	e.log.Debug().Msg("received receipt from requester engine")
+// processEvents is processor of pending events which drives events from networking layer to business logic in `Core`.
+// Effectively consumes messages from networking layer and dispatches them into corresponding sinks which are connected with `Core`.
+// Should be run as a separate goroutine.
+func (e *Engine) processEvents() {
+	// takes pending event from one of the queues
+	// nil sink means nothing to send, this prevents blocking on select
+	fetchEvent := func() (*Event, EventSink, *deque.Deque) {
+		if val, ok := e.pendingReceipts.Front(); ok {
+			return val.(*Event), e.receiptSink, &e.pendingReceipts
+		}
+		if val, ok := e.pendingApprovals.Front(); ok {
+			return val.(*Event), e.approvalSink, &e.pendingApprovals
+		}
+		if val, ok := e.pendingApprovalResponses.Front(); ok {
+			return val.(*Event), e.approvalResponseSink, &e.pendingApprovalResponses
+		}
```

Comment:
I would switch those two cases. If we have requested approvals, they are likely blocking progress. We should process them with priority.